bleak/backends/bluezdbus/agent.py

```diff
@@ -0,0 +1,210 @@
+from bleak.backends.bluezdbus.utils import assert_reply
+from bleak.backends.device import BLEDevice
+
+logger = logging.getLogger(__name__)
+
+
+class Agent(ServiceInterface):
+    """
+    Implementation of the org.bluez.Agent1 D-Bus interface.
+    """
+
+    def __init__(self, callbacks: BaseBleakAgentCallbacks):
+        """
+        Args:
+        """
+        super().__init__(defs.AGENT_INTERFACE)
+        self._callbacks = callbacks
+        self._tasks: Set[asyncio.Task] = set()
+
+    @staticmethod
+    async def _create_ble_device(device_path: str) -> BLEDevice:
+        manager = await get_global_bluez_manager()
+        return BLEDevice(
+            manager.get_device_address(device_path),
+            manager.get_device_name(device_path),
+            {"path": device_path},
+        )
+
+    @method()
+    def Release(self):  # noqa: N802
+        logger.debug("Release")
+
+    # REVISIT: mypy is broke, so we have to add redundant @no_type_check
+    # https://github.com/python/mypy/issues/6583
+
+    @method()
+    @no_type_check
+    async def RequestPinCode(self, device: "o") -> "s":  # noqa: F821 N802
+        logger.debug("RequestPinCode %s", device)
+        raise NotImplementedError
+
+    @method()
+    @no_type_check
+    async def DisplayPinCode(self, device: "o", pincode: "s"):  # noqa: F821 N802
+        logger.debug("DisplayPinCode %s %s", device, pincode)
+        raise NotImplementedError
+
+    @method()
+    @no_type_check
+    async def RequestPasskey(self, device: "o") -> "u":  # noqa: F821 N802
+        logger.debug("RequestPasskey %s", device)
+
+        ble_device = await self._create_ble_device(device)
+
+        task = asyncio.create_task(self._callbacks.request_passkey(ble_device))
+        self._tasks.add(task)
+
+        try:
+            key = await task
+        except asyncio.CancelledError:
+            raise DBusError("org.bluez.Error.Canceled", "task canceled")
+        finally:
+            self._tasks.remove(task)
+
+        if not key:
+            raise DBusError("org.bluez.Error.Rejected", "user rejected")
+
+        try:
+            passkey = int(key)
+        except ValueError:
+            raise DBusError("org.bluez.Error.Rejected", "invalid passkey")
+
+        return passkey
+
+    @method()
+    @no_type_check
+    async def DisplayPasskey(  # noqa: N802
+        self, device: "o", passkey: "u", entered: "q"  # noqa: F821
+    ):
+        passkey = f"{passkey:06}"
+        logger.debug("DisplayPasskey %s %s %d", device, passkey, entered)
+        raise NotImplementedError
+
+    @method()
+    @no_type_check
+    async def RequestConfirmation(self, device: "o", passkey: "u"):  # noqa: F821 N802
+        passkey = f"{passkey:06}"
+        logger.debug("RequestConfirmation %s %s", device, passkey)
+
+        ble_device = await self._create_ble_device(device)
+
+        task = asyncio.create_task(self._callbacks.confirm_passkey(ble_device, passkey))
```

Comment:
Could use a comment explaining that we have to create a task so that it can be canceled by a different callback.